block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagFileStorage.scala

```diff
@@ -137,26 +155,54 @@ final class BlockDagFileStorage[F[_]: Concurrent: Sync: Log: RaiseIOError] priva
       } else if (sortOffset - startBlockNumber + topoSortVector.length < Int.MaxValue) { // Max Vector length
         lock.withPermit(
           for {
-            checkpoints          <- getCheckpoints
-            checkpointsWithIndex = checkpoints.zipWithIndex
-            checkpointsToLoad    = checkpointsWithIndex.filter(startBlockNumber < _._1.end)
-            checkpointsDagInfos <- checkpointsToLoad.traverse {
-                                    case (startingCheckpoint, index) =>
-                                      loadCheckpointDagInfo(startingCheckpoint, index)
-                                  }
-            topoSortPrefix = checkpointsDagInfos.toVector.flatMap { checkpointsDagInfo =>
-              val offset = startBlockNumber - checkpointsDagInfo.sortOffset
-              // offset is always a valid Int since the method result's length was validated before
-              checkpointsDagInfo.topoSort.drop(offset.toInt) // negative drops are ignored
-            }
-            result = topoSortPrefix ++ topoSortVector
+            topoSortPrefix <- topoSortBeforeOffset(startBlockNumber)
+            result         = topoSortPrefix ++ topoSortVector
```

Comment:
Why not move the lock also to `topoSortBeforeOffset`. This should not be responsibility of the caller to care about the lock.